### PR TITLE
Refactor dns server to pass configurables to handlers in a struct.  

### LINKED ISSUE #1332: DNS responses incomplete when TCP query contains EDNS record
```
$ weave launch
$ weave run --name=foo -ti ubuntu
e14c663b3e5996ac94288a947d5985ff89450ffcdc35e04fde23ca25f2f27d8f
$ for i in $(seq 50); do weave dns-add 10.2.1.$i foo -h bar.weave.local ; done
$ dig +short @$(weave docker-bridge-ip) bar.weave.local A | wc -l
50
$ dig +tcp +short @$(weave docker-bridge-ip) bar.weave.local A | wc -l
50
$ dig +bufsize=700 +short @$(weave docker-bridge-ip) bar.weave.local A | wc -l
22
$ dig +tcp +bufsize=700 +short @$(weave docker-bridge-ip) bar.weave.local A | wc -l
21
```

Looking at this a bit more deeply, `dig` by default adds an EDNS OPT RR record indicating the max buffer size is 4k. The `+bufsize` option shrinks that. dig also falls back to tcp when getting a truncated response.

The reason we get truncation even then is the nameserver takes into account the EDNS-supplied UDP buffer size even when interacting over TCP.

I am pretty sure this is wrong.

### PR DESCRIPTION
**Pull request by tomwilkie**
Also ignore EDNS on tcp requests.  Fixes #1332

### CODE REVIEW

File: nameserver/dns.go [hunk 1]
```diff
@@ -286,9 +289,16 @@ func truncateResponse(response *dns.Msg, maxSize int) {
 	response.Truncated = true
 }
 
-func getMaxResponseSize(req *dns.Msg, defaultMaxResponseSize int) int {
+func (h *handler) responseTooBig(request, response *dns.Msg) bool {
+	if len(response.Answer) <= 1 || h.maxResponseSize <= 0 {
```
Comment (tomwilkie) [inline nameserver/dns.go]: This is the bug fix; checkout maxResponseSize > 0 before deciding if a response is too big.

File: nameserver/dns.go [hunk 2]
```diff
@@ -112,144 +112,146 @@ func (d *DNSServer) errorResponse(r *dns.Msg, code int, w dns.ResponseWriter) {
...
+		reqCopy := req.Copy()
+		reqCopy.Id = dns.Id()
+		response, _, err := h.client.Exchange(reqCopy, fmt.Sprintf("%s:%s", server, h.upstream.Port))
+		if err != nil || response == nil {
+			h.ns.debugf("error trying %s: %v", server, err)
+			continue
+		}
+		h.ns.debugf("response: %+v", response)
+		response.Id = req.Id
+		if h.responseTooBig(req, response) {
```
Comment (tomwilkie) [inline nameserver/dns.go]: Note this also fixes the case where recursive dns entries would be considered too big and compressed when using TCP.

File: nameserver/dns.go [hunk 3]
```diff
@@ -70,14 +70,14 @@ func (d *DNSServer) listen(address string) error {
 	if err != nil {
 		return err
 	}
-	udpServer := &dns.Server{PacketConn: udpListener, Handler: d.createMux(d.udpClient, minUDPSize)}
+	udpServer := &dns.Server{PacketConn: udpListener, Handler: d.createMux(d.udpClient, minUDPSize).mux}
```
Comment (rade) [inline nameserver/dns.go]: The `.mux` is a wart. I'd lift the mux creation out of `createMux`, and make that function, suitably renamed, just add the handlers.

File: test/295_dns_large_response_test.sh [hunk 4]
```diff
@@ -16,8 +16,18 @@ for i in $(seq $N); do
...
+
+M=$(exec_on $HOST1 c0 dig +tcp +short $NAME A | grep -v ';;' | wc -l)
+assert_raises "test $M -eq $N"
+
+M=$(exec_on $HOST1 c0 dig +bufsize=700 +short $NAME A | grep -v ';;' | wc -l)
+assert_raises "test $M -eq $N"
+
+M=$(exec_on $HOST1 c0 dig +tcp +bufsize=700 +short $NAME A | grep -v ';;' | wc -l)
+assert_raises "test $M -eq $N"
+
```
Comment (rade) [inline test/295_dns_large_response_test.sh]: A helper function would do wonders here, reducing all the above to
```
check
check +tcp
check      +bufsize=700
check +tcp +bufsize=700
```

File: nameserver/dns.go (hunk 3 above)
Comment (tomwilkie) [inline nameserver/dns.go]: I changed the mux to an embedded struct, so the handler is now a ServeMux.  What do you think?

**Comment by rade** [comment]
> I changed the mux to an embedded struct, so the handler is now a ServeMux. What do you think?

It's not what I would have done. Handlers don't need to know about the ServeMux. Let alone do they need to be *be* a ServeMux. It's unnecessary coupling.

**Comment by rade** [comment]
i.e.
```
func (d *DNSServer) addHandlers(mux *dns.ServeMux, client *dns.Client, defaultMaxResponseSize int) {
	h := &handler{
		DNSServer:       d,
		maxResponseSize: defaultMaxResponseSize,
		client:          client,
	}
	mux.HandleFunc(d.domain, h.handleLocal)
	mux.HandleFunc(reverseDNSdomain, h.handleReverse)
	mux.HandleFunc(topDomain, h.handleRecursive)
}
```
and
```
mux := dns.NewServeMux()
d.addHandlers(mux, d.udpClient, minUDPSize)
udpServer := &dns.Server{PacketConn: udpListener, Handler: mux}
```

**Comment by rade** [comment]
or
```
udpServer := &dns.Server{PacketConn: udpListener, Handler: dns.NewServeMux()}
d.addHandlers(udpServer.Handler, d.udpClient, minUDPSize)
```

File: test/295_dns_large_response_test.sh [hunk 5]
```diff
@@ -16,8 +16,16 @@ for i in $(seq $N); do
 done
 weave_on $HOST1 dns-add $IPS $CID -h $NAME
 
-assert_dns_record $HOST1 c0 $NAME $IPS
-
 assert_raises "exec_on $HOST1 c0 dig MX $NAME | grep -q 'status: NXDOMAIN'"
 
+check() {
+    M=$(exec_on $HOST1 c0 dig +short $@ $NAME A | grep -v ';;' | wc -l)
+    assert_raises "test $M -eq $N"
```
Comment (rade) [inline test/295_dns_large_response_test.sh]: couldn't you shorten that to
```
assert "exec_on ..." $N
```
?

File: nameserver/dns.go [hunk 6]
```diff
@@ -286,9 +291,16 @@ func truncateResponse(response *dns.Msg, maxSize int) {
 	response.Truncated = true
 }
 
-func getMaxResponseSize(req *dns.Msg, defaultMaxResponseSize int) int {
+func (h *handler) responseTooBig(request, response *dns.Msg) bool {
+	if len(response.Answer) <= 1 || h.maxResponseSize <= 0 {
+		return false
+	}
+	return response.Len() > h.getMaxResponseSize(request)
+}
```
Comment (rade) [inline nameserver/dns.go]: That can be just
```
        return h.maxResponseSize > 0 && len(response.Answer) > 1 && response.Len() > h.getMaxResponseSize(request)
```

**Comment by tomwilkie** [comment]
Superseded #1344

**Pull request closed without merging**